cluster/replication/shard_replication_apply.go

```diff
@@ -64,6 +64,34 @@ func (s *ShardReplicationFSM) writeOpIntoFSM(op ShardReplicationOp, status Shard
 		return ErrShardAlreadyReplicating
 	}
 
+	if existingOps, ok := s.opsBySourceFQDN[op.SourceShard]; ok {
+		for _, existingOp := range existingOps {
+			// First check the status of the existing op. If it's READY or CANCELLED we can accept a new op
+			// If it's ongoing we need to check if it's a move, in which case we can't accept any new op
+			// Otherwise we can accept a copy if the existing op is also a copy
+			if existingOpStatus, ok := s.statusById[existingOp.ID]; !ok {
+				// This should never happen
+				return fmt.Errorf("could not find op status for op %d", existingOp.ID)
+			} else if existingOpStatus.GetCurrentState() == api.CANCELLED {
+				continue
+			} else if existingOpStatus.GetCurrentState() == api.READY && existingOp.TransferType == api.COPY {
+				continue
+			}
+
+			// If any of the ops we're handling is a move we can't accept any new op
+			if existingOp.TransferType == api.MOVE {
+				return ErrShardAlreadyReplicating
```

Comment:
As it stands, will this end up being communicated to the user as a `500` at the API layer? Can we catch it there and return a `4xx` error, perhaps `409 CONFLICT`?

Reply:
Ah this was already an issue outside of this PR given L64. Improving the API surface here can come in a separate PR

Reply:
We currently use `strings.Contains` in `func (s *Raft) GetReplicationDetailsByReplicationId` to detect not found errors so could use the same for `func (s *Raft) ReplicationReplicateReplica`. I do wonder if there's a cleaner solution though 🤔